pkgs/dart_mcp/lib/src/api/tools.dart

```diff
@@ -288,6 +331,313 @@ extension type Schema.fromMap(Map<String, Object?> _value) {
+  if (schemaType != null) {
+    switch (schemaType) {
+      case JsonType.object:
+        if (data is Map<String, Object?>) {
+          failures.addAll(_validateObject(schema as ObjectSchema, data));
+        } else {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.list:
+        if (data is List) {
+          failures.addAll(_validateList(schema as ListSchema, data));
+        } else {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.string:
+        if (data is String) {
+          failures.addAll(_validateString(schema as StringSchema, data));
+        } else {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.num:
+        if (data is num) {
+          failures.addAll(_validateNumber(schema as NumberSchema, data));
+        } else {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.int:
+        if (data is int) {
+          failures.addAll(_validateInteger(schema as IntegerSchema, data));
+        } else if (data is num && data == data.toInt()) {
+          failures.addAll(
+            _validateInteger(schema as IntegerSchema, data.toInt()),
+          );
+        } else {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.bool:
+        if (data is! bool) {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.nil:
+        if (data != null) {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+    }
+  }
+
+  return failures.toSet().toList(); // Remove duplicates
+}
+
+List<ValidationFailureType> _validateObject(
+  ObjectSchema schema,
+  Map<String, Object?> data,
+) {
+  final failures = <ValidationFailureType>[];
+
+  if (schema.minProperties != null &&
+      data.keys.length < schema.minProperties!) {
+    failures.add(ValidationFailureType.minPropertiesNotMet);
+  }
+  if (schema.maxProperties != null &&
+      data.keys.length > schema.maxProperties!) {
+    failures.add(ValidationFailureType.maxPropertiesExceeded);
+  }
+
+  for (final reqProp in schema.required ?? const []) {
+    if (!data.containsKey(reqProp)) {
+      failures.add(ValidationFailureType.requiredPropertyMissing);
+    }
+  }
+
+  final evaluatedKeys = <String>{};
+  if (schema.properties != null) {
+    for (final entry in schema.properties!.entries) {
+      if (data.containsKey(entry.key)) {
+        evaluatedKeys.add(entry.key);
+        failures.addAll(
+          _validateSchema(
+            entry.value,
+            data[entry.key],
+          ).map((_) => ValidationFailureType.propertyValueInvalid),
+        );
+      }
+    }
+  }
+  if (schema.patternProperties != null) {
+    for (final entry in schema.patternProperties!.entries) {
```

Comment:
I think some comments around these sections might be good, to explain what is being checked exactly (same for the list validator).